drivers/adc/adc_shell.c

```diff
@@ -52,178 +72,139 @@ struct adc_hdl {
+	{ .string = "ADC_GAIN_1_4", .gain = ADC_GAIN_1_4 },
+	{ .string = "ADC_GAIN_1_3", .gain = ADC_GAIN_1_3 },
+	{ .string = "ADC_GAIN_1_2", .gain = ADC_GAIN_1_2 },
+	{ .string = "ADC_GAIN_2_3", .gain = ADC_GAIN_2_3 },
+	{ .string = "ADC_GAIN_1", .gain = ADC_GAIN_1 },
+	{ .string = "ADC_GAIN_2", .gain = ADC_GAIN_2 },
+	{ .string = "ADC_GAIN_3", .gain = ADC_GAIN_3 },
+	{ .string = "ADC_GAIN_4", .gain = ADC_GAIN_4 },
+	{ .string = "ADC_GAIN_8", .gain = ADC_GAIN_8 },
+	{ .string = "ADC_GAIN_16", .gain = ADC_GAIN_16 },
+	{ .string = "ADC_GAIN_32", .gain = ADC_GAIN_32 },
+	{ .string = "ADC_GAIN_64", .gain = ADC_GAIN_64 }
 };
 
-/** get_adc_from_list returns the number entry of the adc in the adc_list,
- * returns -ENODEV if it doesn't exist
- */
-static int get_adc_from_list(char *name)
-{
-	int adc_idx;
+static const struct reference_string_to_enum {
+	char *string;
+	enum adc_reference reference;
+} reference_list[] = {
+	{ .string = "VDD_1", .reference = ADC_REF_VDD_1 },
+	{ .string = "VDD_1_2", .reference = ADC_REF_VDD_1_2 },
+	{ .string = "VDD_1_3", .reference = ADC_REF_VDD_1_3 },
+	{ .string = "VDD_1_4", .reference = ADC_REF_VDD_1_4 },
+	{ .string = "INT", .reference = ADC_REF_INTERNAL },
+	{ .string = "EXT0", .reference = ADC_REF_EXTERNAL0 },
+	{ .string = "EXT1", .reference = ADC_REF_EXTERNAL1 }
+};
 
-	for (adc_idx = 0; adc_idx < ARRAY_SIZE(adc_list); adc_idx++) {
-		if (!strcmp(name, adc_list[adc_idx].device_name)) {
-			return adc_idx;
+static struct adc_hdl *get_adc(const char *device_label)
+{
+	for (int i = 0; i < ARRAY_SIZE(adc_list); i++) {
+		if (!strcmp(device_label, adc_list[i].device_label)) {
+			return &adc_list[i];
 		}
 	}
-	return -ENODEV;
+
+	/* This will never happen because ADC was prompted by shell */
+	__ASSERT_NO_MSG(false);
+	return NULL;
 }
 
 static int cmd_adc_channel(const struct shell *shell, size_t argc, char **argv)
 {
-	int retval = 0;
+	/* -1: index of ADC label name */
+	struct adc_hdl *adc = get_adc(argv[-1]);
 	const struct device *adc_dev;
-	int chosen_adc;
-
-	if (argc != args_no.channel) {
-		shell_fprintf(shell, SHELL_NORMAL,
-				"Usage: channel <channel_id>\n");
-		return 0;
-	}
-
-	chosen_adc = get_adc_from_list(argv[args_indx.adc]);
-	if (chosen_adc < 0) {
-		shell_error(shell, "Device not in device list");
-		return -EINVAL;
-	}
+	int retval = 0;
 
-	adc_dev = device_get_binding(adc_list[chosen_adc].device_name);
+	adc_dev = device_get_binding(adc->device_label);
 	if (adc_dev == NULL) {
 		shell_error(shell, "ADC device not found");
 		return -ENODEV;
 	}
+
 	if (!isdigit((unsigned char)argv[args_indx.conf][0])) {
 		shell_error(shell, "<channel> must be digits");
 		return -EINVAL;
 	}
-	adc_list[chosen_adc].channel_config.channel_id =
+
+	adc->channel_config.channel_id =
 		(uint8_t)strtol(argv[args_indx.conf], NULL, 10);
-	retval = adc_channel_setup(adc_dev,
-			&adc_list[chosen_adc].channel_config);
+
+	if (IS_ENABLED(CONFIG_ADC_CONFIGURABLE_INPUTS)) {
+		adc->channel_config.input_positive = 1;
```

Comment:
This should not be hard-coded. Users should be able to select the input to use. And this value may be even invalid for some drivers.

Reply:
@anangl Do you think it is a good idea to add one more command for setting this value?